src/replicache.ts

```diff
@@ -572,23 +591,51 @@ export default class Replicache implements ReadTransaction {
    * mutations.
    */
   async push(): Promise<void> {
-    if (this._pushTimerId) {
-      clearTimeout(this._pushTimerId);
-      this._pushTimerId = 0;
-    }
+    this._clearPushTimer();
     await this._wrapInOnlineCheck(() => this._push(MAX_REAUTH_TRIES), 'Push');
   }
 
   private async _push(maxAuthTries: number): Promise<void> {
+    // Is there already a push in flight?
+    if (this._pushCounter > 0) {
+      // Here we create a promise/resolve pair that will resolve when the
+      // current push is done.
+      //
+      // This whole dance is to collapse incoming calls to push during an in
+      // flight push so that they only cause a single actual push RPC. We also want
+      // the call to push to resolve at the right time.
+      //
+      // - push() is called. Lets call this "pushA"
+      // - pushA is being sent over wire
+      // - push() is called again. Lets call this "pushB"
+      // - push() is called again. Lets call this "pushC"
+      // - pushA comes back and is resolved
+      // - puhsB is being sent over wire
+      // - pushB resolve comes back and is resolved
+      // - pushC resolves
+
+      if (!this._pushResolver) {
```

Comment:
I have trouble following this. Why does the following pattern not work?

```
void async foo() {
  if (_fooPromise != null) {
    await _fooPromise;
    return;
  }
  _fooPromise = _foo();
  await _fooPromise;
  _fooPromise = null;
}

void async _foo() {

}
```

Reply:
That does not deal with the case where push is called during an existing push. In that case it needs to be run after the current push.

Reply:
Then one might think you can just do:

```dart
void async foo() {
  if (_fooPromise != null) {
    await _fooPromise;
    await _foo();
    return;
  }
  _fooPromise = _foo();
  await _fooPromise;
  _fooPromise = null;
}
```

but then you do not get collapsing of multiple calls to push during an outstanding push. I thought the following would work:

```dart
void async foo() {
  if (_fooPromise != null) {
    await _fooPromise;
    if !_nextFooPromise) {
      _nextFooPromise = _foo();
    }
    await _nextFooPromise();
    _nextFooPromise = null;
    return;
  }
  _fooPromise = _foo();
  await _fooPromise;
  _fooPromise = null;
}
```

That hung the tests and I thought I knew why but the reason escaped...

Reply:
> That does not deal with the case where push is called during an existing push. In that case it needs to be run after the current push.

Yes. That is why in my first sketch I had the target iterate until it wasn't dirty anymore.